### PR TITLE
Implement Tagged UNION Type

### PR DESCRIPTION
**Pull request by Maxxen**
Draft PR for #4302

I still feel very unsure of what I'm doing, and I know there's all sorts of crimes going on in `vector_cast.cpp`, but I got a basic implementation of the functionality discussed in the linked issue working. 

Although I've added a new UNION LogicalType it's based on the struct PhysicalType. 

I had to overload `struct_extract` to enable accessing union members but I'm not sure it's the most elegant implementation. We would probably like provide UNION specific error messages.

Some TODOs:
- [ ] Figure out how to use ValidityMask instead of manually nulling the inactive union members.
- [ ] Add remaining `<type>` to Union casts
- [ ] Add reverse Union to `<type>` casts
- [ ] Allow casting a single-member struct to unions to disambiguate the desired member when a union contains multiple members of the same type. 
- [ ] Add auxiliary helper functions utilizing unions. Thinking something like `extract_or(<union>, 'member', <fallback>)`
- [ ] Add lots more tests

### CODE REVIEW

File: src/common/vector_operations/vector_cast.cpp [hunk 1]
```diff
@@ -12,6 +12,43 @@
 #include "duckdb/common/limits.hpp"
 namespace duckdb {
 
+static bool ToUnionCast(Vector &source, Vector &result, idx_t count, string *error_message) {
+	auto &candidates = UnionType::GetChildrenOfType(result.GetType(), source.GetType());
+
+	if(candidates.size() == 0) {
+		throw TypeMismatchException(source.GetType(), result.GetType(), "Cannot cast to UNION");
```
Comment (Mytherin) [inline src/common/vector_operations/vector_cast.cpp:19]: A more descriptive error message would be preferable here.

File: src/common/vector_operations/vector_cast.cpp [hunk 2]
```diff
@@ -12,6 +12,43 @@
...
+	}
+	else if(candidates.size() == 1){
```
Comment (Mytherin) [inline src/common/vector_operations/vector_cast.cpp:21]: Prefer early-out to make code more readable/easier to follow: `if (candidates.size() != 1) { throw ... } `

File: test/sql/types/union/union_test.test [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# name: test/sql/union/union_test.test
+# description: Test union type
+# group: [union]
+
+statement ok
+CREATE TABLE tbl(u UNION(i INTEGER, v VARCHAR));
```
Comment (Mytherin) [inline test/sql/types/union/union_test.test:6]: Some more suggestions for tests:

* Inserting multiple values into a UNION table at once (e.g. `INSERT INTO tbl SELECT i FROM range(10) tbl(i)`)
* Inserting `UNION` values directly (e.g. `CREATE TABLE tbl2 AS SELECT * FROM tbl`, or `INSERT INTO tbl2 SELECT * FROM tbl`, where `tbl` and `tbl2` have UNION columns with multiple values)
* Duplicate name in union entry
* Attempting to extract a non-existent column from a union
* Extracting unions of different vector types (e.g. add a `WHERE` clause to create a dictionary vector)
* Manual casts to union types
* Storing NULL values in union columns
* Unions with multiple identical types (e.g. `UNION(i INTEGER, j INTEGER)`)
* Lists inside unions (e.g. `UNION(i INTEGER[], j INTEGER)`)
* Lists OF unions (e.g. `UNION(i INTEGER, s VARCHAR)[]`)
* Nested unions (union inside of another union)
* `UNION` different `UNION` types (hah) - (i.e. `SELECT 1::UNION(i INTEGER) UNION ALL SELECT 2::UNION(i BIGINT)`).
* GROUP BY/DISTINCT on union columns
* Persistent storage of union types (see the tests in `test/sql/storage`)

File: src/common/vector_operations/vector_cast.cpp [hunk 4]
```diff
@@ -12,6 +12,43 @@
...
+				}
+			}
+		}
+
+		return true;
+	}
+	else {
+		// We have multiple union members of the same type.
+		// TODO: Allow users to disambiguate by passing a struct as a key and value pair
+		throw TypeMismatchException(source.GetType(), result.GetType(), "More than one candidate for UNION cast");
```
Comment (Mytherin) [inline src/common/vector_operations/vector_cast.cpp:47]: Perhaps list the candidates here in the error message?

File: src/common/vector_operations/vector_cast.cpp [hunk 5]
```diff
@@ -12,6 +12,43 @@
...
+		for(idx_t child_idx = 0; child_idx < child_vectors.size(); child_idx++) {
+			auto &child = child_vectors[child_idx];
+			
+			if(child->GetType() == candidate_type) {
+				child->ReferenceAndSetType(source);
+			}
+			else {
+				// TODO: We need to OR with the inverted validity mask of the selected candidate child
+				// This probably wont work..?
+				for(idx_t j = 0; j < count; j++) {
```
Comment (Mytherin) [inline src/common/vector_operations/vector_cast.cpp:36]: We generally turn the vector into a constant null in a case like this, e.g.:

```
result.SetVectorType(VectorType::CONSTANT_VECTOR);
ConstantVector::SetNull(result, true);
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! This is a great start, and a very cool & promising feature! Some comments:

**Comment by Maxxen** [comment]
Thanks a ton for the feedback so far! I haven't prioritised this feature since the draft PR went up but I've let it marinate in the back of my mind and have some more ideas to test out.

Primarily I think we should store the tag itself into the union primary vector while the members are stored in children instead of implicitly have it be determined by the  first non-null member. This will require a more advanced logical type implementation but I think it will be more performant and robust to work with. 

I do feel more confident diving into the internals now though and I'm hoping to pick it up again soon!

**Comment by Mytherin** [comment]
> Primarily I think we should store the tag itself into the union primary vector while the members are stored in children instead of implicitly have it be determined by the first non-null member. This will require a more advanced logical type implementation but I think it will be more performant and robust to work with.

That sounds like a cleaner solution indeed, although that might require creating a new physical type which would create a bit more work on e.g. the storage layer.

**Pull request closed without merging**